pkg/storegateway/bucket_streaming_readers.go

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package storegateway
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storegateway/storegatewaypb"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+	"github.com/grafana/mimir/pkg/util/limiter"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// The code in this file is used by the queriers to read the streaming chunks from the storegateway.
```

Comment:
Why did you choose to have this in the storegateway package? The existing code for the block store querying is in `pkg/querier/blocks_store_queryable.go` and it feels like `SeriesChunksStreamReader` this is doing a very similar thing as the existing code (querying series from the long-term store).

Reply:
It was in querier package before, but since similarly named code exists for ingesters, from Charles' comment, which I agree with, keeping it here makes it easier to read the code when we use it as `storegateway.NewSeriesChunksStreamReader` (and similarly `ingester.NewSeriesChunksStreamReader`). An alternative is to rename all the structs and constructors here to include the component names and move it to querier package.

Reply:
it can also live in a separate package. The `ingester` package that is used for the querier<>ingester streaming is specifically an ingester client.

I think having a separate package somewhat protects against coupling the querier and storegateway packages. For example, this file updates the querier stats. I don't think the storegateway package should be aware of what stats the querier keeps track of.

WDYT about 
* moving this to the storegatewaypb package 
	* So it's used as `storegatewaypb.NewSeriesChunksStreamReader`
* and adding another implementation of this client interface which is only responsible for keeping track of the stats; this will live somewhere in the querier packages
	* I'd even go further and move the limit tracking into its own client